### PR TITLE
core(codegen) the css optimization

### PR DESCRIPTION
**Pull request by Jefiozie**
First version of the CSS optimization is ready

Closes: #136

### CODE REVIEW

File: packages/xlayers/src/app/editor/code-editor/editor-container/codegen/shared-codegen.service.ts [hunk 1]
```diff
@@ -10,7 +11,10 @@ export enum Template {
 })
 export class SharedCodegen {
   private indentationSymbol = '  '; // 2 spaces ftw
-
+  generateComponentOptimizedStyles(ast: SketchMSLayer) {
+    const optimizer = new StyleSheetOptimizer(true);
```
Comment (manekinekko) [inline packages/xlayers/src/app/editor/code-editor/editor-container/codegen/shared-codegen.service.ts]: we should have this as an angular service.

**Comment by manekinekko** [comment]
@Jefiozie can you update the message of your commit to `chore:` ?

File: packages/xlayers/src/app/editor/code-editor/editor-container/codegen/shared-codegen.service.ts (hunk 1 above)
Comment (Jefiozie) [reply packages/xlayers/src/app/editor/code-editor/editor-container/codegen/shared-codegen.service.ts]: I was thinking about that do we really want this. For now I will make the change we can always refactor it.

Comment (manekinekko) [reply packages/xlayers/src/app/editor/code-editor/editor-container/codegen/shared-codegen.service.ts]: this would allow people to easily switch implementation if they like.

**Comment by Jefiozie** [comment]
- [ ] Move the optimizer to the parser package
- [ ] Add some basic documentation
- [ ] Add some edge case tests

**Comment by manekinekko** [comment]
@Jefiozie Awesome! This PR is 98% done. We just need to figure out if it is worth it to run multiple optimizer passes in order to remove as many duplicates as possible.
